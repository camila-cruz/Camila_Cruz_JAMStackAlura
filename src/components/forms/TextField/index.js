/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
  margin-bottom: 32px;
  position: relative;
`;

const Input = styled(Text)`
  width: 100%;
  height: 48px;
  border: 1px solid ${({ theme }) => theme.mainUi.background.light.tertiary};
  border-radius: 10px;
  padding: 14px 6px;
  margin-top: 4px;
  outline: 0;
  
  ${({ textarea }) => textarea && css`height: 80px;`};
  ${({ isFieldInvalid }) => isFieldInvalid && css`
    border-color: ${({ theme }) => theme.actions.error};
  `};

  &:focus {
    box-shadow: 0 0 3pt 1pt ${({ theme }) => theme.mainUi.background.light.tertiary};
  }
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'inputText',
};

export default function TextField({
  placeholder,
  name,
  id,
  type,
  textarea,
  value,
  error,
  isTouched,
  onChange,
  ...props
}) {
  const hasError = Boolean(error);
  const isFieldInvalid = hasError && isTouched;

  return (
    <InputWrapper>
      {(textarea
        && (
          <Input
            tag="textarea"
            name={name}
            id={id}
            value={value}
            onChange={onChange}
            placeholder={placeholder}
            isFieldInvalid={isFieldInvalid}
            rows={10}
            style={{ resize: 'none' }}
            textarea
            {...props}
          />
        )) || (
          <Input
            placeholder={placeholder}
            name={name}
            id={id}
            value={value}
            type={type}
            onChange={onChange}
            isFieldInvalid={isFieldInvalid}
            {...props}
          />
      )}
      {isFieldInvalid && (
        <Text
          variant="errorFormText"
          display="block"
          marginTop="4px"
          action="error"
          role="alert"
        >
          {error}
        </Text>
      )}
    </InputWrapper>
  );
}

TextField.defaultProps = {
  placeholder: null,
  type: 'text',
  value: '',
  textarea: false,
  onChange: null,
  error: '',
  isTouched: false,
};

TextField.propTypes = {
  placeholder: PropTypes.string,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  value: PropTypes.string,
  onChange: PropTypes.func,
  type: PropTypes.string,
  textarea: PropTypes.bool,
  error: PropTypes.string,
  isTouched: PropTypes.bool,
};
