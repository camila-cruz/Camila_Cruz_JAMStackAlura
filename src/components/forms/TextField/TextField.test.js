import React from 'react';
import user from '@testing-library/user-event';
import TextField from '.';
import { render, screen } from '../../../infra/test/testUtils';

describe('<TextField />', () => {
  test('renders a textarea', () => {
    render(
      <TextField
        textarea
        value=""
        placeholder="Deixe uma mensagem"
        name="message"
        id="message"
        onChange={() => {}}
      />,
    );

    const textareaField = screen.getByPlaceholderText('Deixe uma mensagem');
    expect(textareaField).toMatchSnapshot();
  });

  describe('when the field is valid', () => {
    describe('and the user types', () => {
      test('the value should be updated', () => {
        const onSubmitMock = jest.fn();
        const fieldValueMock = 'Jequiti';

        render(
          <TextField
            placeholder="Nome"
            value=""
            name="name"
            id="nome"
            onChange={onSubmitMock}
          />,
        );

        const nameField = screen.getByPlaceholderText(/nome/i);
        user.type(nameField, fieldValueMock);

        expect(onSubmitMock).toHaveBeenCalledTimes(fieldValueMock.length);
      });
    });
  });

  describe('when the field is invalid', () => {
    describe('and untouched', () => {
      test('no errors should be displayed', () => {
        render(
          <TextField
            placeholder="Nome"
            value=""
            name="name"
            id="nome"
            onChange={() => {}}
          />,
        );

        const nameField = screen.getByPlaceholderText(/nome/i);
        expect(nameField).toMatchSnapshot();
      });
    });

    describe('and has been touched', () => {
      test('an error should be displayed below it', () => {
        const fieldValueMock = 'emailaleatorio.com';
        const errorMock = 'O campo e-mail deve ter um valor v√°lido';

        render(
          <TextField
            type="email"
            placeholder="E-mail"
            value={fieldValueMock}
            name="email"
            id="email"
            isTouched
            error={errorMock}
            onChange={() => {}}
          />,
        );

        const emailField = screen.getByPlaceholderText(/e-mail/i);
        expect(emailField).toHaveValue(fieldValueMock);

        const errorMessage = screen.getByRole('alert');
        expect(errorMessage).toHaveTextContent(errorMock);

        expect(emailField).toMatchSnapshot();
      });
    });
  });
});
