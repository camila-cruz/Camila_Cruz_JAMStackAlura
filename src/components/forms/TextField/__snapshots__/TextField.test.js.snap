// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders a textarea 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 600;
  line-height: normal;
}

.c1 {
  width: 100%;
  height: 48px;
  border: 1px solid #512919;
  border-radius: 10px;
  padding: 14px 6px;
  margin-top: 4px;
  outline: 0;
  height: 80px;
}

.c1:focus {
  box-shadow: 0 0 3pt 1pt #512919;
}

<textarea
  class="c0 c1"
  id="message"
  name="message"
  placeholder="Deixe uma mensagem"
  rows="10"
  style="resize: none;"
/>
`;

exports[`<TextField /> when the field is invalid and has been touched an error should be displayed below it 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 600;
  line-height: normal;
}

.c1 {
  width: 100%;
  height: 48px;
  border: 1px solid #512919;
  border-radius: 10px;
  padding: 14px 6px;
  margin-top: 4px;
  outline: 0;
  border-color: #d63031;
}

.c1:focus {
  box-shadow: 0 0 3pt 1pt #512919;
}

<input
  class="c0 c1"
  id="email"
  name="email"
  placeholder="E-mail"
  type="email"
  value="emailaleatorio.com"
/>
`;

exports[`<TextField /> when the field is invalid and untouched no errors should be displayed 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 600;
  line-height: normal;
}

.c1 {
  width: 100%;
  height: 48px;
  border: 1px solid #512919;
  border-radius: 10px;
  padding: 14px 6px;
  margin-top: 4px;
  outline: 0;
}

.c1:focus {
  box-shadow: 0 0 3pt 1pt #512919;
}

<input
  class="c0 c1"
  id="nome"
  name="name"
  placeholder="Nome"
  type="text"
  value=""
/>
`;
